sciencesearch/nlp/visualize_kws.py

```diff
@@ -0,0 +1,245 @@
+        for phrase in keywords:
+            tokenized_phrase = word_tokenize(phrase.lower())
+            stemmed_phrase = tuple(self.caseSS.stem(word) for word in tokenized_phrase)
+            self.stemmed_kw_set.add(stemmed_phrase)
+        
+        self.max_kw_length = max([len(word_tokenize(phrase)) for phrase in keywords], default=1)
+
+
+    def style_text(self, tokens, ngram):
+        if ngram in self.stemmed_kw_set:
+            styled_str = f"<span class='{self.class_name}'>{' '.join(tokens)}</span> "
+            return True, styled_str
+        else:
+            return False, ""
+
+
+class MultiSet_Visualizer(KWS_Visualizer):
+
+    def __init__(self, keywords_dict: dict[str, list[str]], text: str = None, txt_filepath = None):
+        super().__init__(text=text, txt_filepath=txt_filepath)
+        self.keywords_dict = keywords_dict
+        
+        # Process each set of keywords
+        self.stemmed_kw_sets = {}
+        max_length = 1
+
+        for set_name, keywords in keywords_dict.items():
+            self.stemmed_kw_sets[set_name] = set()
+            for phrase in keywords:
+                phrase_tokens = word_tokenize(phrase.lower())
+                stemmed_phrase = tuple(self.caseSS.stem(token) for token in phrase_tokens)
+                self.stemmed_kw_sets[set_name].add(stemmed_phrase)
+                max_length = max(max_length, len(phrase_tokens))
+        
+        self.max_kw_length = max_length
+    
+
+    def style_text(self, tokens, ngram):
+        # Determine which sets this n-gram belongs to
+        matching_sets = []
+        for set_name, stemmed_set in self.stemmed_kw_sets.items():
+            if ngram in stemmed_set:
+                matching_sets.append(set_name)
+        
+        if matching_sets:
+            # Create a class name based on which sets matched
+            class_name = " ".join([f"kw-{set_name}" for set_name in matching_sets])
+            styled_str = f'<span class="{class_name}">{" ".join(tokens)}</span> '
+            return True, styled_str
+        return False, ""
+
+
+class HTMLBuilder:
+
+    def __init__(self, visualizer: KWS_Visualizer, filename: str, title: str, 
+                 css_styles: dict[str, str] = None):
+        self.visualizer = visualizer
+        self.filename = filename
+        self.title = title
+        self.html = ""
+        self.css_styles = css_styles or {}
+
+    def get_highlighted_html(self):
+        html_body =  self.visualizer.highlight_tokens_html()
+        html_output = self.generate_html(self.title, html_body)
+        self.html = html_output
+
+
+    def write_file_and_run(self):
+        file = open(self.filename,'w')
+        file.write(self.html) 
+        file.close() 
+        fp = 'file:///'+os.getcwd()+'/' + self.filename
+        webbrowser.open_new_tab(fp) 
+
+    def generate_html(self, title: str = "" ,body_content: str = ""):
+
+        legend = ""
+        if isinstance(self.visualizer, MultiSet_Visualizer):
```

Comment:
consider adding a get_legend() method to the Visualizer classes instead of testing subclass (fragile)